src/app/CommandHandler.h

```diff
@@ -190,9 +226,73 @@ class CommandHandler : public Messaging::ExchangeDelegate
     CHIP_ERROR AddClusterSpecificFailure(const ConcreteCommandPath & aCommandPath, ClusterStatus aClusterStatus);
 
     Protocols::InteractionModel::Status ProcessInvokeRequest(System::PacketBufferHandle && payload, bool isTimedInvoke);
-    CHIP_ERROR PrepareCommand(const ConcreteCommandPath & aCommandPath, bool aStartDataStruct = true);
+
+    /**
+     * This adds a new CommandDataIB element into InvokeResponses for the associated
+     * aRequestCommandPath. This adds up until the `CommandFields` element within
+     * `CommandDataIB`.
+     *
+     * This call will fail if CommandHandler is already in the middle of building a
+     * CommandStatusIB or CommandDataID (ei something has called Perpare*, without
+     * calling Finish*), or is already sending InvokeResponseMessage.
+     *
+     * Upon success, the caller is expected to call `FinishCommand` once they have added
+     * all Data into Fields element of CommandDataIB.
+     *
+     * @param [in] aResponseCommandPath the concrete response path that we are sending to Requester.
+     * @param [in] aPrepareParameters struct containing paramters needs for preparing a command, such as request path.
+     */
+    CHIP_ERROR PrepareInvokeResponseCommand(const ConcreteCommandPath & aResponseCommandPath,
+                                            const InvokeResponseParameters & aPrepareParameters);
+
+    [[deprecated("PrepareCommand now needs the requested command path. Please use PrepareInvokeResponseCommand")]] CHIP_ERROR
+    PrepareCommand(const ConcreteCommandPath & aCommandPath, bool aStartDataStruct = true);
+
+    /**
+     * Finishes the CommandDataIB element within the InvokeResponses.
+     *
+     * Caller must have first successfully called `PrepareInvokeResponseCommand`.
+     *
+     * @param [in] aEndDataStruct end the TLV container for the CommandFields element within
+     *             CommandDataIB.
+     *
+     * @return CHIP_ERROR_INCORRECT_STATE
+     *                      If device has not previously successfully called
+     *                      `PrepareInvokeResponseCommand`.
+     * @retval CHIP_ERROR_BUFFER_TOO_SMALL
```

Comment:
So when this error happens... what should the caller do?  Is there an error status in the buffer now?  Or does one need to be written?  Is it safe to touch the buffer at all?  It's going to get auto-sent, so ideally it would be in a "good" state....

Followup to sort this out.

Reply:
I mean as of today we have the rollback/checkpoint stuff that is "supposed" send just a status update. I say supposed to in quote because we know that doesn't always happen. I will be fixing this in followup PRs. I don't think there is a need to mention it again as I have already added one TODO comment about it